app/grandchallenge/reader_studies/forms.py

```diff
@@ -310,10 +336,38 @@ def interface_choices(self):
     def initial_interface(self):
         return self.interface_choices().first()
 
+    def widget_choices(self):
+        answer_type = self["answer_type"].value()
+        choices = [("", "Default")]
+        if answer_type:
+            try:
+                extra_options = ANSWER_TYPE_TO_QUESTION_WIDGET[answer_type]
+                choices.extend(
+                    (option.name, option.label) for option in extra_options
+                )
+            except KeyError:
+                pass
+        return choices
+
+    def initial_widget(self):
+        try:
+            return self.instance.widget.kind
+        except ObjectDoesNotExist:
+            return ""
+
+    def clean_widget(self):
+        value = self.cleaned_data.get("widget")
+        if value == "":
+            return None
+        else:
+            return QuestionWidget(kind=value)
```

Comment:
The `full_clean()` method will fail here because the `question` is not set. We can't set the question at this point, because then we can no longer check and remove the old widget in the save method.

Reply:
Can you track the old widget in the model? Override the `__init__` method and add something like `self._widget_orig`, then check after `super().save(...)` in `save`?

Reply:
Actually, I wonder if we need a separate form all-together for the Widget as it seems we're having to hack around this a lot. Maybe an [inline form set](https://docs.djangoproject.com/en/3.2/topics/forms/modelforms/#inline-formsets) is what we need, with a separate form for the widget?

Reply:
Or, maybe we should be updating the existing widget rather than creating and deleting?

Reply:
Or maybe we should give up on this design and do everything with a JSON field on the Question model...

Reply:
I will think some more and try some stuff.

Reply:
Setting `self._widget_orig` doesn't help us here. The `full_clean` method would still fail if we create a new widget for a question that already has a widget, because you can only have one widget associated with a particular quesiton. So for the full_clean method to work we would need to in fact delete the old widget before we call it, which would mean deleting the object in the `clean` method

Reply:
I'm looking into formsets now. They might be a good option, especially if we want to add options to the widget model in the future, but I still need to work out how to update the widget choices within the formset depending on answer_type.

Reply:
Formsets don't seem easier to work with than what we have here in terms of validation or updating the choices per answer_type, so I'm hesitant to move towards using formsets. 

Here is a summary of the options I think we have: 
1. Use the current implementation, and do `full_clean(exclude="question")` in the form's `clean` method. That defeats the purpose of calling `full_clean` though. 
2. Use current implementation and only call the `full_clean` method on the final model instance in the form's `save` method - also not great since that's not what the `save` method should be used for. 
3. Do what we currently do for attaching default answers to a question. Users have to contact support for a new widget, which we create in the admin and they can choose from a list of filtered options. Relatively simple to implement in the form (using a dynamic ModelChoiceField), and since we'd only have one single widget for now, also no extra admin overhead for the time being. 
4. Drop the `QuestionWidget` model entirely and instead use a json field on the `Question` model to define the widget. We would define a json schema, for now super simple, but can be made as complex as we need it to be in the future. Gives the user more autonomy than option 3, but if we end up with lots of configuration options, validation gets complex.

Reply:
I tend towards option 3 right now

Reply:
I don't think that widgets should be shared across questions, too easy to change and then you affect multiple questions at once. The workflow of asking for a widget from support does not scale with options, so when that happens the same issues that we face now would need to somehow be solved.

I really think that changing the model is the way to go, how about:

- `question.widget = models.CharField(choices=QuestionWidgetChoices, ...)`

Then in the future we add `question.widget_options` as a JSON field. It's really not that different to the approach we're going for right now.

Reply:
Ok, so then we get rid of the `QuestionWidget` model and only add a `widget` choice field on the `Question` model for now.

Reply:
Yes

Reply:
Probably should have gone for that when we realised that multiple concrete models were not feasible.

Reply:
Probably, I didn't foresee the issues it would cause with validation on the form, my bad. I think it's probably best if I close this PR and start a new one, right?

Reply:
Ah it's OK, my fault too, at least we got to a good design before it's used. Up to you whether you want to create a new one or re-work this.